### ISSUE TITLE
Wrong virtual display size with different resolutions

### ISSUE THREAD

**Issue by huy-hng** [wheaney/decky-XRGaming#34]
Game Tested: Hades

When changing the resolution to something that is lower than the native resolution of the xr glasses (viture pro in my case) then all modes except VR-lite result in the display being too small.

I assume this is game independent, therefore reproduction steps are as follows:

1. Set in game resolution to something lower than the native glasses resolution

2. Set mode to either virtual display or follow (the latter should have smooth follow enabled)

**Comment by wheaney** [comment]
It's not really a bug. The screen is _actually_ smaller when it's lower resolution. You can zoom it in from the plugin controls. I would rather the user explicitly choose to do this than do it automatically, because there's an impact to the decision to fit the display to the XR glasses resolution that I don't want to assume the user is okay with (namely that it's a lower resolution image and will be a bit blurry when zoomed in). 

So your options are:
1. Zoom it in
2. Have it use the native glasses' resolution either in the game settings or by using one of the scaling options in the performance sidebar (e.g. FSR).

**Comment by huy-hng** [comment]
Oh I forgot to mention that it's smaller than even that.
Setting the resolution to 1080p fills out the screen as expected.
Setting it to 800p or 720p shrinks the screen to be smaller (if integer scaling is selected in the steam overlay)

If I enable the plugin the screen becomes even smaller than the 800p naturally smaller screen. This happens for any scaling mode.
One can also see some scaling artifacts in the image, which isn't there when disabling the plugin.

You can test this by lowering the in-game resolution and then switching between Follow and disabled or setting the mode to follow and toggling smooth follow.

Hope this helps!

**Comment by cool0221** [comment]
Wanted to pick up this thread and describe the issues more. I installed the plugin on my steam deck today. Under setting/display I enabled automatic resolutions. I tried out a few games where I set the resolutions to native under each games’ setting menu. The result is mixed. Both Black Myth Wukong and Shiro Coaltown worked flawlessly, with the virtual display having exactly the same screen size versus the original size when XR is toggled off. Also the picture quality is superb with minimal visible noise; however, for red dead redemption 2 (steam version) as well as the Google Chrome (steam shortcut added from desktop install), both virtual displays appear a lot smaller than the original size. Zoom in helps with the size, but the picture is a lot noisier, also the mouse pointer doesn’t work on the zoomed in screen anymore. I hope this helps you to look into it more. Thanks for such a great product!

**Comment by cool0221** [comment]
Just to be clear, for all the trials above I set the display size slider always to 1.0 in the plugin.

**Comment by cool0221** [comment]
Also - for Wukong and Shiro Coaltown, if I set the game resolution to default under each game’s setting, I will also get smaller and noisy virtual display. Changing to native helps to resolve this issue. But not for the other two games so it seems.

**Comment by KristofPusztai** [comment]
I've also noticed this for virtual display, the display size is very small despite setting game resolution to native in games (in this case "Ghost of Tsushima") settings.

**Comment by wheaney** [comment]
Let me say one thing up front: the behavior you're seeing here is the result of the game's resolution. You've seen that when a game has the correct resolution (1080p) it will be full screen at a size/distance of 1.0, which demonstrates that it operates correctly when the resolution is correct. Breezy/XRGaming doesn't control the application's resolution, so this issue can't really go anywhere.

With that said, there are several factors in determining the application's resolution. This isn't an either/or thing, all three have to be true simultaneously:
1. SteamOS -- in the `Settings` -> `Displays` tab -- has to see that the display can do 1080p, and it must choose that display resolution. You can hope it's doing this if you choose the `Automatic` resolution, or you can explicitly choose `1920x1080` from the options.
2. SteamOS must launch the game/app in a compositor view that is 1080p. This requires that before you launch the game (when you see the green `Play` button), you go into the game properties, find the resolution menu, and choose `Native` or explicitly choose the 1080p option. If you're not seeing the 1920x1080 option, then Step 1 hasn't been completed adequately.
3. After you launch the game/app, go into that games graphics/video settings, and set the resolution to 1920x1080. If you're not seeing that option, then steps 1 and/or 2 have not been completed adequately. 
    * For an app like Chrome, I'm not sure how it determines the resolution to launch at, though I would expect it to use full-screen. Alternatively, you can set the app's resolution lower and use a scaling option -- like FSR -- to upscale it to 1080p, but this may result in fuzziness/blurring/lower-quality.

I'll re-iterate: Breezy/XRGaming doesn't control the application's resolution. So the above is the best advice I can offer. Any other troubleshooting with the resolution is out of scope for this app. So I'm going to close this out.

**Comment by KristofPusztai** [comment]
Yup, I’m also experiencing the same thing that @lonsdaleite is describing, even with FSR kicking in which should be upscaling to the 1080p resolution so there shouldn't be any scale down occurring?

**Comment by wheaney** [comment]
@lonsdaleite @KristofPusztai The bad news is that everything I said above is still true. [Here's the line in my shader](https://github.com/wheaney/sombrero/blob/76cb6bb65e90dfe2abf5bd2db8e280ae6e03d5c6/Sombrero.frag#L425) where I grab the source (game) resolution via a couple variables within the reshade namespace, as provided by the vulkan layer on the steam deck. So what you're seeing, unfortunately, is that the deck is running my shader on the content at the original resolution (e.g. 720p), and then applying FSR afterwards. Ideally it would apply my stuff *after* FSR so my effect would alter the upscaled image, but I have no control over how Steam Deck is implementing its reshade shader support.

**Comment by lonsdaleite** [comment]
@wheaney If I set the game to 720p and adjust display size to fill the screen — am I still getting proper FSR upscale? Or does the order of operations result in something different than expected? Just want to make sure the quality isn’t degraded.

**Comment by wheaney** [comment]
> am I still getting proper FSR upscale?

Probably not. If my effect is modifying the 720p image, then it'll cause that image to lose some detail (due to the downscaling you're seeing, plus impacts of cropping and viewing angle) and then the anchored view (including the black background and everything that results from that view) is scaled up to 1080p. Assuming all that is true, 720p plus FSR will produce a significantly worse result than rendering straight to 1080p and avoiding the upscaling.
